### PR TITLE
Add noise to perfect mutation matrix

### PR DESCRIPTION
**Pull request by gordonkoehn**

### CODE REVIEW

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 1]
```diff
@@ -44,15 +47,58 @@ def add_noise_to_perfect_matrix(
...
-    rng_false_pos, rng_false_neg, rng_miss = random.split(rng, 3)
-
-    # TODO(Pawel, Gordon): Now probably it's the easiest to draw from random.bernoulli
-    #   matrices corresponding to the false positives and false negatives
-    #   and adjust the mutation matrix accordingly
-
-    # Now the matrix can be adjusted according to missing entry simulation
-
-    raise NotImplementedError("This function needs to be implemented.")
+    rng_false_pos, rng_false_neg, rng_miss, rng_homo_pos, rng_homo_neg = random.split(
```
Comment (pawel-czyz) [inline src/pyggdrasil/tree_inference/_simulate.py:50]: Very nice you did it! However, I'd encourage you to split the body of the function into several smaller functions (a function with 50 lines of code, multiple steps, involving complex constructions may be hard to understand in a month or two).

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 2]
```diff
@@ -44,15 +47,58 @@ def add_noise_to_perfect_matrix(
...
+    )
+    # make matrix to edit and keep unchanged
+    noisy_mat = matrix.copy()
+
+    # Add False Positives - P(D_{ij} = 1 |E_{ij}=0)=alpha
+    # Generate a random matrix of the same shape as the original
+    rand_matrix = random.uniform(key=rng_false_pos, shape=matrix.shape)
+    # Create a mask of elements that satisfy the condition
+    # (original value equals y and random value is less than p)
+    mask = (matrix == 0) & (rand_matrix < false_positive_rate)
```
Comment (pawel-czyz) [inline src/pyggdrasil/tree_inference/_simulate.py:61]: Cool, trick. For the future, you may also consider [jax.random.choice](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.choice.html)

File: tests/tree_inference/test_simulate.py [hunk 3]
```diff
@@ -1 +1,276 @@
...
+        false_positive_rate,
+        false_negative_rate,
+        missing_entry_rate,
+        observe_homozygous,
+    )
+
+    freq_fp = np.sum((noisy_mat == 1) & (perfect_mat == 0)) / (n * m)
+    #
+    # # # three standard deviations - stdev = (N * p * (1-p))^0.5
+    # tolerance = (
```
Comment (pawel-czyz) [inline tests/tree_inference/test_simulate.py:125]: The next six lines look like commented out code, so it'd be good to remove it.

(Comments should contain only human-readable information + instructions for the code linters. Commented-out code gets outdated pretty quickly.)

**Review by pawel-czyz — APPROVED** [review]
Nice work! I'd suggest to remove the commented out code (for reasons given in the appropriate comment) and perhaps (not that important) to split the long function body into several smaller steps.

Please, merge after the comments are resolved :slightly_smiling_face:

File: tests/tree_inference/test_simulate.py (hunk 3 above)
Comment (gordonkoehn) [reply tests/tree_inference/test_simulate.py:125]: OK removed - makes sense. This was my attempt to find a statistically meaningful tolerance. It was too rigorous for some reason, hence I reverted to absolute fixed tolerance of 3% frequency.

File: src/pyggdrasil/tree_inference/_simulate.py (hunk 2 above)
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_simulate.py:61]: Yes, thank you - discovered this too late.

File: src/pyggdrasil/tree_inference/_simulate.py (hunk 1 above)
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_simulate.py:186]: Alright, split add_noise_to_perfect_matrix() into four private functions.

**Review by gordonkoehn — COMMENTED** [review]
Feedback implemented: Commented Code Removed + Function fractioned for readability

**Review by gordonkoehn — COMMENTED** [review]
Feedback implemented: Comments removed + add_noise...() function split into parts

**Pull request merged**